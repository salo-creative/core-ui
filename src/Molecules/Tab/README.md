# Tab

A component to wrap other components in tabs.

---

## Usage

Install

```javascript
yarn add @salo/core-ui
```

Include

```javascript
import Tab from ' @salo/core-ui/Molecules/Tab';
```

Implement as follows

```javascript
const panes = [
  { 
    title: 'Tab 1', 
    render: (
      <p>This tab has JSX content</p>
    ),
  }
];

<Tab panes={ panes } />
```

Or with complex titles:

```javascript
const panes = [
  { 
    title: 'Tab 1',
    children: <h1>Tab 1</h1>,
    render: (
      <p>This tab has JSX content</p>
    ),
  }
];