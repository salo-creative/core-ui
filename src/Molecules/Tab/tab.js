import React from 'react';
import PropTypes from 'prop-types';
import { memoize } from 'lodash';

// COMPONENTS & STYLES
import { TabButton, Pane } from './tab.styles';

function computeMachineName(name) {
  return name.toLowerCase().replace(/ /g, '-');
}

const getMachineName = memoize(computeMachineName);

const Tab = (props) => {
  const { className, panes } = props;
  const [activeIndex, setActiveIndex] = React.useState(0);
  const [focus, setFocus] = React.useState(0);

  const handleDirectionalArrows = (event) => {
    let tabFocus = focus;
    if (event.keyCode === 39 || event.keyCode === 37) {
      if (event.keyCode === 39) {
        tabFocus++;
        // If we're at the end, go to the start
        if (tabFocus >= panes.length) {
          tabFocus = 0;
        }
        // Move left
      } else if (event.keyCode === 37) {
        tabFocus--;
        // If we're at the start, go to the end
        if (tabFocus < 0) {
          tabFocus = panes.length - 1;
        }
      }
      setFocus(tabFocus);
      document.querySelector(`#tab-${ getMachineName(panes[tabFocus].title) }`).focus();
    }
  };

  return (
    <div className={ `${ className } salo-tabs` }>
      <div
        aria-label='Tabs'
        className='salo-tabs__tabs'
        onKeyDown={ handleDirectionalArrows }
        role='tablist'
        tabIndex={ 0 }
      >
        { panes.map((pane, index) => {
          return (
            <TabButton
              aria-controls={ getMachineName(pane.title) }
              aria-selected={ index === activeIndex }
              className='salo-tabs__button'
              id={ `tab-${ getMachineName(pane.title) }` }
              key={ pane.title }
              onClick={ () => setActiveIndex(index) }
              role='tab'
              tabIndex={ index === focus ? 0 : -1 }
              type='button'
            >
              { pane.children || pane.title }
            </TabButton>
          );
        }) }
      </div>
      <div>
        { panes.map((pane, index) => (
          <Pane
            aria-labelledby={ `tab-${ getMachineName(pane.title) }` }
            className='salo-tabs__pane'
            hidden={ index !== activeIndex }
            id={ getMachineName(pane.title) }
            key={ getMachineName(pane.title) }
            role='tabpanel'
            tabIndex={ index === focus ? 0 : -1 }
          >
            { index === activeIndex && pane.render() }
          </Pane>
        )) }
      </div>
    </div>
  );
};

Tab.defaultProps = {
  className: ''
};

Tab.propTypes = {
  className: PropTypes.string,
  panes: PropTypes.arrayOf(
    PropTypes.shape({
      children: PropTypes.any,
      title: PropTypes.string,
      render: PropTypes.func.isRequired
    })
  ).isRequired
};

export default Tab;