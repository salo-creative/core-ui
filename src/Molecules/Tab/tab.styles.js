import styled from 'styled-components';

export const TabButton = styled.button`
  appearance: none;
  border-radius: 5px 5px 0 0;
  border: 0;
  font-size: 1.4rem;
  font-weight: 700;
  padding: 15px 30px;
  position: relative;

  /* Specific styles if selected */
  ${ (props) => {
    if (props['aria-selected']) {
      return `
        background-color: ${ props.theme.paleGrey };
        box-shadow: 0px -1px 1px 0px ${ props.theme.grey };
      `;
    }
    return '';
  } }

  &:focus {
    box-shadow: 0px -2px 2px 0px ${ ({ theme }) => theme.grey };
  }

  &:active {
    top: 1px;
  }
`;

export const Pane = styled.div`
  background-color: ${ ({ theme }) => theme.paleGrey };
  border-radius: 0 0 5px 5px;
  box-shadow: 1px 2px 1px 0px ${ ({ theme }) => theme.grey };
  padding: 30px;
`;