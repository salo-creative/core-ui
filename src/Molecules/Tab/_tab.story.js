import React from 'react';
import { withKnobs } from '@storybook/addon-knobs';

// load tests
import { withTests } from '@storybook/addon-jest';
import results from '../../../.storybook/jest-test-results.json';

// FEATURED COMPONENT
import { Tab } from '../../index';

// README //
import README from './README.md';

// Start of story logic
export const Basic = () => {
  return (
    <Tab
      panes={ [
        {
          title: 'Sawtooth',
          render: () => {
            return (
              <p>
                The sawtooth wave (or saw wave) is a kind of non-sinusoidal waveform. It is so named
                based on its resemblance to the teeth of a plain-toothed saw with a zero rake angle.
              </p>
            );
          }
        },
        {
          title: 'Sine',
          render: () => {
            return (
              <p>
                A sine wave or sinusoid is a mathematical curve that describes a smooth periodic
                oscillation.
              </p>
            );
          }
        },
        {
          title: 'Triangle',
          render: () => {
            return (
              <React.Fragment>
                <p>
                  A triangle wave is a non-sinusoidal waveform named for its triangular shape. It is
                  a periodic, piecewise linear, continuous real function.
                </p>
                <img
                  src='https://upload.wikimedia.org/wikipedia/commons/4/4f/Triangle-td_and_fd.png'
                  alt='A bandlimited triangle wave pictured in the time domain (top) and frequency domain (bottom). The fundamental is at 220 Hz (A3).'
                />
              </React.Fragment>
            );
          }
        }
      ] }
    />
  );
};

Basic.story = {
  decorators: [
    withKnobs,
    withTests({
      results
    })
  ],
  parameters: {
    jest: ['tab'],
    notes: README
  }
};

export default {
  title: 'Molecules/Tab'
};