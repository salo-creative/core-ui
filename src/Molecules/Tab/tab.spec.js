import React from 'react';
import 'jest-styled-components';

import { renderWithTheme } from '../../../test';

// COMPONENT
import Tab from './index';

test('Renders against snapshot', async () => {
  const panes = [{
    title: 'Item 1',
    render: () => 'Item 1 content'
  }];
  const { container } = renderWithTheme(
    <Tab panes={ panes } />
  );
  expect(container.firstChild).toMatchSnapshot();
});