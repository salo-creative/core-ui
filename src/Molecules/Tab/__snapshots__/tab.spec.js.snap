// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders against snapshot 1`] = `
.c0 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  border-radius: 5px 5px 0 0;
  border: 0;
  font-size: 1.4rem;
  font-weight: 700;
  padding: 15px 30px;
  position: relative;
  background-color: #fafcff;
  box-shadow: 0px -1px 1px 0px #e9ecef;
}

.c0:focus {
  box-shadow: 0px -2px 2px 0px #e9ecef;
}

.c0:active {
  top: 1px;
}

.c1 {
  background-color: #fafcff;
  border-radius: 0 0 5px 5px;
  box-shadow: 1px 2px 1px 0px #e9ecef;
  padding: 30px;
}

<div
  class=" salo-tabs"
>
  <div
    aria-label="Tabs"
    class="salo-tabs__tabs"
    role="tablist"
    tabindex="0"
  >
    <button
      aria-controls="item-1"
      aria-selected="true"
      class="c0 salo-tabs__button"
      id="tab-item-1"
      role="tab"
      tabindex="0"
      type="button"
    >
      Item 1
    </button>
  </div>
  <div>
    <div
      aria-labelledby="tab-item-1"
      class="c1 salo-tabs__pane"
      id="item-1"
      role="tabpanel"
      tabindex="0"
    >
      Item 1 content
    </div>
  </div>
</div>
`;
